webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -204,89 +382,117 @@ trait PromiseGenerationService extends Loggable {
+      /// only keep successfull node config. We will keep the failed one to fail the whole process in the end if needed
+      nodeConfigs           =  allNodeConfigs.ok.map(c => (c.nodeInfo.id, c)).toMap
       timeBuildConfig       =  (System.currentTimeMillis - buildConfigTime)
-      _                     =  logger.debug(s"Node's target configuration built in ${timeBuildConfig} ms, start to update rule values.")
+      _                     =  PolicyLogger.debug(s"Node's target configuration built in ${timeBuildConfig} ms, start to update rule values.")
 
       updatedNodeConfigIds  =  getNodesConfigVersion(nodeConfigs, nodeConfigCaches, generationTime)
       updatedNodeConfigs    =  nodeConfigs.filterKeys(id => updatedNodeConfigIds.keySet.contains(id))
 
       reportTime            =  System.currentTimeMillis
       expectedReports       =  computeExpectedReports(ruleVals, updatedNodeConfigs.values.toSeq, updatedNodeConfigIds, generationTime, allNodeModes)
       timeSetExpectedReport =  (System.currentTimeMillis - reportTime)
-      _                     =  logger.debug(s"Reports updated in ${timeSetExpectedReport} ms")
+      _                     =  PolicyLogger.debug(s"Reports updated in ${timeSetExpectedReport} ms")
 
       ///// so now we have everything for each updated nodes, we can start writing node policies and then expected reports
 
-      // WHY DO WE NEED TO FORGET OTHER NODES CACHE INFO HERE ?
-      _                     <- forgetOtherNodeConfigurationState(nodeConfigs.keySet) ?~! "Cannot clean the configuration cache"
+      // we want to remove cache for nodes for which we didn't successfully get a config (redo next time)
+      // and also remove all non active nodes (deleted, etc) => only keep the one for which we have a node configuration
+      _                     <- forgetOtherNodeConfigurationState(allNodeConfigs.ok.map(_.nodeInfo.id).toSet) ?~! "Cannot clean the configuration cache"
+
+      // clear cache of string template - both before and after writing them
+      techniqueResources    <- getTechniquesResources(updatedNodeConfigIds.keySet, nodeConfigs)
+      _                     =  clearCacheOfTemplate()
+
 
+      //////  write new configurations
       writeTime             =  System.currentTimeMillis
-      writtenNodeConfigs    <- writeNodeConfigurations(rootNodeId, updatedNodeConfigIds, nodeConfigs, allLicenses, globalPolicyMode, generationTime) ?~!"Cannot write nodes configuration"
-      timeWriteNodeConfig   =  (System.currentTimeMillis - writeTime)
-      _                     =  logger.debug(s"Node configuration written in ${timeWriteNodeConfig} ms, start to update expected reports.")
+      remainingNodes        =  AtomicInt(updatedNodeConfigIds.size)
+      allTrySavedReports    =  ParallelSequence.applicative(updatedNodeConfigIds.toSeq) { case (nodeId, nodeConfigId) =>
```

Comment:
you need to first generate for policy servers, and then for nodes

Reply:
done